### ISSUE TITLE
📋Create guide step validation system

### ISSUE THREAD

**Issue by CalvinWilkinson** [KinsonDigital/VelaptorDocs#129]
### Complete The Item Below

- [X] I have updated the title without removing the 📋 emoji.

### Description

Create a step validation system.  This will prevent the problem of publishing the website with steps that do not exist or are physically out of order.

> **Note**
> Create a VSCode launch config to be used for debugging.

## Step Validation Status Check Requirements

This explains the requirements for the status check script that will be used to validate the ***steps***
that exist in the various guides in the Velaptor docs website.

***What does "step validation" mean?***

The problem is that keeping track of the order of the steps in the guides is tedious and error prone.  When creating new guides, or editing existing ones, it is common to have the steps not be in order, or not be numbered sequentially.  This makes this unclear and confusing for the user.

***What is the goal?***

The goal is to have the script executed by a GitHub workflow.  This GitHub workflow will be a status check that will run
every time a change occurs with a pull request.  If the status check fails, this means there is an issue with one or more
of the steps in a particular file.

### **Website Step Example:**
<img width="196" alt="image" src="https://github.com/KinsonDigital/VelaptorDocs/assets/85414302/857211d9-63e1-42f0-b026-31bba027412f">

### Validation Requirements

Guide steps are made up of a main step or a sub step.  A sub step can exist as a "child" step of a main step.

In docusaurus, steps are automatically detected based on the use of the heading markdown syntax.

#### **Standard Markdown Docusaurus Step Examples:**

For a main step, if there is a line of markdown that contains 2 x `#` symbol, then it is a main step.
```md

## This is a main step
```

For a sub-step, if there is a line of markdown that contains 3 x `#` symbol, then it is a sub-step.
```md

### This is a sub-step
```

The text of a main or sub-step can be literally anything.  This follows standard markdown standards.  For the Velaptor
doc website, a main step consists of a custom ***React*** component where the contents of the component HTML is the text that
will be displayed on the right side of the website.

#### **Velaptor Docs Step Examples:**

**Main Step**
```md
## <span class="color-step">Step 1: Project setup</span>
```

**Sub-Step**
```md
### <span class="color-sub-step">Step 2.2: Creating a mouse object</span>
```

The React component is 👉🏻 ```<span class="color-sub-step"></span>```

You will notice that the React component contains a class attribute with the name `color-step` or `color-sub-step`.  This
changes the color of the step.  Main steps are a darker brown, and sub-steps are a lighter brown. This is standard CSS.

The content of a step would be 👉🏻 Step 2.2: `Creating a mouse object`

This can be anything, but the Velaptor website follows a standard.  The standard is always the word 'Step' followed by a space, then a whole number of floating point number with only a single decimal place of precision.

If it is a whole number, then it is a main step, and floating point numbers are sub-steps.

#### **Validating Steps**

This will be done by checking that the following are true:

1. Line does not start with a space.
2. Line starts with only `##` or `###` hash symbols.
3. The hashes and the HTML (React component) are separated by a single space.
4. The `class` attribute must be a value of `color-step` or `color-sub-step`.
5. The HTML content must start with either `Step <number>` or `Step <number.number>`.
6. After the `Step <number>` content section, there must be a single `:` symbol
7. After the step number with the colon, there must be come textual content.
    > **Note**
    > The textual content can be anything.  It only has to exist.

When checking the syntax of the steps, this will determine whether or not the steps are a main step or a sub-step.

#### **Validating Step Existence**

This is the concept of checking that a particular step with a number exists in the group of main steps or a group of
sub-steps.

**Example:**
If we had only main steps in the document, and the steps were 1-4, we need to validate that steps 2 and 3 exist.
If we only had steps 1,2, and 4, then this is not valid.

This same concept would be applied to sub-steps.  The different is that we are checking for sub-steps, not main steps.

**Valid Example:**

```md
## <span class="color-step">Step 1</span>
## <span class="color-step">Step 2</span>
## <span class="color-step">Step 3</span>
## <span class="color-step">Step 4</span>
```

**Invalid Example:**

```md
## <span class="color-step">Step 1</span>
## <span class="color-step">Step 2</span>
## <span class="color-step">Step 4</span>
```

### **Validating Step Order**

This is the concept of making sure that the steps are physically in order in the document.  It would not make sense to have
4 main steps, and have the order of the steps from top to bottom out of order.

**Valid Example:**

```md
## <span class="color-step">Step 1</span>
## <span class="color-step">Step 2</span>
## <span class="color-step">Step 3</span>
```

**Invalid Example:**

```md
## <span class="color-step">Step 2</span>
## <span class="color-step">Step 1</span>
## <span class="color-step">Step 3</span>
```

#### **Other Considerations**

1. The validation of step existence should be checked **BEFORE** checking for the step order.  The reason is because, if we are missing steps, then step order does not matter.  Validating the concept or step order depends on the existence of step existence
in the first place.
1. Getting a list of files that are only markdown files are going to be required.
    > **Note**
    > This is because we need to know where the files are and the fully qualified file paths before we can load the file.
1. The loading of file content is going to be required for.
1. The script is going to have to know the difference between a main step and a sub-step
1. We need to print the problems that we find to the GitHub console, so that way the user knows how to fix the issue.

### Acceptance Criteria

```[tasklist]
### Validation
- [ ] Step validation script created
- [ ] Step existence validation added
- [ ] Step order validation added
``` 

```[tasklist]
### Problem console messages
- [ ] Step does not exist
- [ ] Step out of order
- [ ] Step content missing `Step <number>` numbering
- [ ] Step textual content missing
```

```[tasklist]
### CICD/Config
- [ ] Status check workflow created
- [ ] VSCode launch task created
```

### ToDo Items

- [X] Change type labels added to this issue.  Refer to the _**Change Type Labels**_ section below.
- [X] Priority label added to this issue.  Refer to the _**Priority Type Labels**_ section below.

### Additional Information:

**_<details closed><summary>Change Type Labels</summary>_**

| Change Type  | Label                                                                                             |
|--------------|---------------------------------------------------------------------------------------------------|
| CICD                 | https://github.com/KinsonDigital/VelaptorDocs/labels/%E2%99%BB%EF%B8%8Fcicd               |
| Dependency Update    | https://github.com/KinsonDigital/VelaptorDocs/labels/%F0%9F%93%A6dependency%20update                  |
| Miscellaneous        | https://github.com/KinsonDigital/VelaptorDocs/labels/miscellaneous                        |
| Tutorial Doc Changes | https://github.com/KinsonDigital/VelaptorDocs/labels/%F0%9F%93%9Ctutorial%20doc%20changes |
</details>

**_<details closed><summary>Priority Type Labels</summary>_**

| Priority Type       | Label                                                                  |
|---------------------|------------------------------------------------------------------------|
| Low Priority        | https://github.com/KinsonDigital/VelaptorDocs/labels/low%20priority    |
| Medium Priority     | https://github.com/KinsonDigital/VelaptorDocs/labels/medium%20priority |
| High Priority       | https://github.com/KinsonDigital/VelaptorDocs/labels/high%20priority   |
</details>

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct.
